### PR TITLE
TS 4263: Global key block configurable via Records.config

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci]

File: mgmt/RecordsConfig.cc [hunk 1]
```diff
@@ -1245,7 +1245,7 @@ static const RecordElement RecordsConfig[] =
   ,
   {RECT_CONFIG, "proxy.config.ssl.server.multicert.exit_on_load_fail", RECD_INT, "0", RECU_RESTART_TS, RR_NULL, RECC_NULL, "[0-1]", RECA_NULL}
   ,
-  {RECT_CONFIG, "proxy.config.ssl.server.ticket_key.filename", RECD_STRING, "ssl_ticket.key", RECU_DYNAMIC, RR_NULL, RECC_NULL, NULL, RECA_NULL}
+  {RECT_CONFIG, "proxy.config.ssl.server.ticket_key.filename", RECD_STRING, "ticket.key", RECU_DYNAMIC, RR_NULL, RECC_NULL, NULL, RECA_NULL}
```
Comment (jpeach) [inline mgmt/RecordsConfig.cc]: Please update the documentation, ``doc/admin-guide/files/records.config.en.rst``.

Comment (zwoop) [inline mgmt/RecordsConfig.cc]: I'm still not sure that it's worthwhile renaming it, the old name was pretty descriptive :). But I don't feel strongly.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/642/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/538/ for details.

File: iocore/net/SSLUtils.cc [hunk 2]
```diff
@@ -2018,11 +2047,17 @@ SSLParseCertificateConfiguration(const SSLConfigParams *params, SSLCertLookup *l
...
+  ats_scoped_str ticket_key_filename;
   unsigned line_num = 0;
   matcher_line line_info;
 
   const matcher_tags sslCertTags = {NULL, NULL, NULL, NULL, NULL, NULL, false};
 
+  // load the global ticket key for later use
+  REC_ReadConfigStringAlloc(ticket_key_filename, "proxy.config.ssl.server.ticket_key.filename");
+  ats_scoped_str ticket_key_path(Layout::relative_to(params->serverCertPathOnly, ticket_key_filename));
+  global_default_keyblock = ssl_create_ticket_keyblock(ticket_key_path); // this function just returns a keyblock
```
Comment (zwoop) [inline iocore/net/SSLUtils.cc]: Question: What happens if the ticket file is empty? You create an empty key block, and then try to look up in that and fail (hopefully) gracefully?

That does seem reasonable, but I wonder if the logic should be to do the old default, and look for the dest_ip=*'s keychain at that point? At a minimum, we'd want to document in the release notes that going forward with 7.0.0, someone (like us) using the dest_ip=* as a default now must move that to the global configuration.

Alternatively, maybe we should have a warning that if someone has enabled session tickets, but not providing a key block for either all lines in ssl_multicert.config *or* at least one global one in tickets.config? That seems a little complicated to deal with though, so maybe that's a future RFE.

Comment (persiaAziz) [inline iocore/net/SSLUtils.cc]: hmm.. I have a checking for the file path, if the file path is null, then a random key will be generated. I guess we could do the same for invalid files too?

**Comment by persiaAziz** [comment]
If the ticket key file has a keyblock which is less than 48 bytes, then the assertion statement in ssl_callback_session_ticket(..) will fail. This is the current logic. One thing that confuses is that we create random ticket keys if the filename of key is NULL but fail on assertion if the keyblock is NULL. Can't we make a new random keyblock in such case? @shinrich @jpeach @zwoop

**Comment by shinrich** [comment]
I think we should fail, if the user gives us a bad file (file to short or file not present).  In that case it seems that they intended to specify a key but failed somehow.

If they don't specify any file, then they are wanting to go with a default of creating a random key on the file.

**Pull request merged**